Refactor version handling during build and release

tomaas-zeman opened the pull request:
Based on the previous discussion https://github.com/Heegu-sama/Homm3BG/pull/553#discussion_r1626067929

- version is stored in `.version` file and not parsed from the document
- version number is then injected during build
- ~~versions are also replaced in `README` links during release~~

DarthGandalf commented on tools/release.sh:
```diff
@@ -42,4 +42,10 @@ echo "Optimizing economy printable build..."
+echo "Updating README links..."
+sed -E -i.tmp "s;Version [.0-9]+;Version ${VERSION};" README.md
+sed -E -i.tmp "s;releases/download/v[^/]+;releases/download/v${VERSION};" README.md
+sed -E -i.tmp "s;Rules_Rewrite_[0-9]+;Rules_Rewrite_${FILE_VERSION};" README.md
```
Make it `1.2`?

I imagine we can potentially go above version 9, which would be 10, and eventually have the same name

qwrtln replied:
Initially, we avoided dots in the file names. Do all systems support it?

tomaas-zeman replied:
Code removed

DarthGandalf commented on metadata.tex:
```diff
@@ -53,6 +54,13 @@
+% \getenv[\macro]{variable_name} stores the value in \macro for reusability
+\newcommand{\getenv}[2][]{%
+  \CatchFileEdef{\value}{"|kpsewhich --var-value #2"}{\endlinechar=-1}%
+  \if\relax\detokenize{#1}\relax\value\else\let#1\value\fi}
```
Why do you need this over simple `\input{.version}`?

tomaas-zeman replied:
That's good. I didn't know about that

The `\getenv` macro might be needed for this though, unless input also reads env variables

- https://github.com/Heegu-sama/Homm3BG/actions/runs/9379614635/job/25824911861
- https://github.com/Heegu-sama/Homm3BG/pull/555/commits/122d86e576ea5e486cf35a185e47cdd5accbc71c

DarthGandalf replied:
Currently environment is read in perl in latexmkrc

tomaas-zeman replied:
Can I define any variable in the latexmkrc or do I have to squueze everything into the `pre_tex_code`?

tomaas-zeman replied:
Maybe there's an easier way (shellesc package, latexmkrc) but I was not able to make anything else work.

But I'm open to ideas :)

tomaas-zeman replied:
I changed the `kpsewhich` command to simple `echo` to get the env variable. That should be more acceptable :)

qwrtln replied:
A controversial question: will it work on Windows?

tomaas-zeman replied:
Does it have to? I mean all scripts are written in bash :)

qwrtln replied:
The owner doesn't work actively on this any more, but up until now you could just open the project in TeXStudio and build over there. I'm afraid that with this change it will become impossible.

qwrtln replied:
OMG, it breaks at getenv, not here haha.

tomaas-zeman replied:
Anyway, I'll follow your recommendations anyway on this topic but supporting Windows is just another burden to bear. So unless there's a good reason to do so (like at least few active devs), I would discourage that

tomaas-zeman replied:
So if I go back to the previous command, I still don't have a way to test that it works properly. I could assume it'll work but still

qwrtln commented:
This will not work without refactoring the workflow.

DarthGandalf commented on metadata.tex:
```diff
@@ -53,6 +54,13 @@
+% \getenv{variable_name} just prints the value
+% \getenv[\macro]{variable_name} stores the value in \macro for reusability
+\newcommand{\getenv}[2][]{%
+  \CatchFileEdef{\value}{"|kpsewhich --var-value #2"}{\endlinechar=-1}%
```
Sorry, why do you need to run external command (part of kpathsea) for this?

tomaas-zeman replied:
Sorry, I'm not latex pro and I didn't find a better way to read env variables

DarthGandalf replied:
Don't worry, me neither :D

tomaas-zeman replied:
Let's continue here https://github.com/Heegu-sama/Homm3BG/pull/555/files#r1627353364

qwrtln commented:
> Based on the previous discussion [#553 (comment)](https://github.com/Heegu-sama/Homm3BG/pull/553#discussion_r1626067929)
> 
>     * versions are also replaced in `README` links during release


I'm not sure if it's necessary. Readme evolves every time there's a release. We're not at the point where we need automation here. If anything, it should be a separate script, because I don't want to edit README when I build the release files. And you run it once for every language. And no, I don't want it to build every language on a single run, as a simple bash for loop will do. I want the capability to build a single language.

qwrtln commented on main_en.tex:
```diff
@@ -8,9 +8,7 @@
-\newcommand{\version}{
-  Version 1.2
-}
+\newcommand{\versionlabel}{Version}
```
```suggestion
\newcommand{\version}{
  Version  \input{.version}
}
```
Something simple like this will not do?

tomaas-zeman replied:
done

qwrtln commented on .github/workflows/build-and-publish-rule-book.yaml:
```diff
@@ -25,51 +25,31 @@ jobs:
       - name: Set up repository
         uses: actions/checkout@v4
 
-      - name: Replace rule book version with a warning
```
Why? We want this.

qwrtln replied:
Ah, ok, you put it in the latex code.

DarthGandalf replied:
Thanks, I was going to move that warning over as well, to have all the strings (which are not in .po) translated in the same place

tomaas-zeman commented:
> > Based on the previous discussion [#553 (comment)](https://github.com/Heegu-sama/Homm3BG/pull/553#discussion_r1626067929)
> > ```
> > * versions are also replaced in `README` links during release
> > ```
> 
> I'm not sure if it's necessary. Readme evolves every time there's a release. We're not at the point where we need automation here. If anything, it should be a separate script, because I don't want to edit README when I build the release files. And you run it once for every language. And no, I don't want it to build every language on a single run, as a simple bash for loop will do. I want the capability to build a single language.



> > Based on the previous discussion [#553 (comment)](https://github.com/Heegu-sama/Homm3BG/pull/553#discussion_r1626067929)
> > ```
> > * versions are also replaced in `README` links during release
> > ```
> 
> I'm not sure if it's necessary. Readme evolves every time there's a release. We're not at the point where we need automation here. If anything, it should be a separate script, because I don't want to edit README when I build the release files. And you run it once for every language. And no, I don't want it to build every language on a single run, as a simple bash for loop will do. I want the capability to build a single language.

OK. I'll revert that change

qwrtln commented on .version:
```diff
@@ -0,0 +1 @@
+1.2
```
I guess this file needs to be without the final newline in order not to confuse `cat` or whatever else?

tomaas-zeman replied:
How can I exclude it from the check?

qwrtln replied:
Did you check this? I just tried it locally. Works even with the final newline.

tomaas-zeman replied:
Hmm, it does. That's quite odd. Maybe it would make sense to me in case of `\input` macro but certainly not for `cat`

tomaas-zeman replied:
TIL: "Command substitutions strip all trailing newlines from the output of the command inside them"

https://mywiki.wooledge.org/CommandSubstitution

qwrtln replied:
TIL too :D

qwrtln commented on metadata.tex:
```diff
@@ -278,7 +289,13 @@
 
 \begin{center}
-  \version{}
+  \getenv[\githubsha]{GITHUB_SHA}
```
Gosh, I had no idea LaTeX had such capabilities. I'm a little uneasy with adding an extra package just for this, but on the other hand I like the parametrization, which simplifies the workflow.

qwrtln replied:
How about putting this under another iftoggle, which kicks in only on GitHub? It should allow us to bypass the TeXstudio error?

tomaas-zeman replied:
That could work. How to detect the github env?

qwrtln replied:
Just define another `iftoggle`, and pass it over in the `pre_compile:` arg of the LaTeX GitHub action.

qwrtln replied:
Something like `githubbild`?

qwrtln replied:
Exactly the same way that `printable` is passed, but make use it on GitHub only.

qwrtln commented on .github/workflows/build-and-publish-rule-book.yaml:
```diff
@@ -10,9 +10,9 @@ on:
-  REPO_URL: https://github.com/Heegu-sama/Homm3BG
   GHOSTSCRIPT_URL: https://github.com/ArtifexSoftware/ghostpdl-downloads/releases/download/gs10030/gs_10.03.0_amd64_snap.tgz
   ARTIFACT_RETENTION_DAYS: 1
+  HOMM3_GITHUB_BUILD: 1
```
Didn't work. Please try putting it in the `pre_complie:` arg in every step which builds the PDF file.

tomaas-zeman replied:
Yeah, I noticed. I kind of expected this would fail but still wanted to give it a try and avoid unnecessary duplication

qwrtln replied:
Confirmed 👍🏻

qwrtln commented:
@csagataj2 You were the one? Please share some good news with us and tell us if it builds.

tomaas-zeman commented:
@qwrtln It seems that job config change also changed the structure of checks and the deprecated one are not running now, yet they still seem to be required

<img width="859" alt="Screenshot 2024-06-05 at 15 36 09" src="https://github.com/Heegu-sama/Homm3BG/assets/1928291/ca166baf-a40a-4f47-9a4c-c17296087c43">

qwrtln commented:
> @qwrtln It seems that job config change also changed the structure of checks and the deprecated ones are not running now, yet they still seem to be required
> <img alt="Screenshot 2024-06-05 at 15 36 09" width="859" src="https://private-user-images.githubusercontent.com/1928291/336859130-ca166baf-a40a-4f47-9a4c-c17296087c43.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc1OTk1MzksIm5iZiI6MTcxNzU5OTIzOSwicGF0aCI6Ii8xOTI4MjkxLzMzNjg1OTEzMC1jYTE2NmJhZi1hNDBhLTRmNDctOWE0Yy1jMTcyOTYwODdjNDMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MDVUMTQ1MzU5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YTNmOGMxYzc5OTA4YWJiYTNlNTFkNmEyZGQ4ZWIwY2FhOWI5ZjA2NjI3YWUzMWU4YTY5NDVkNDk1MmRkMGUwMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.cUWeEjY3tUTRArLywvgWkltwdcF72cYBAF3WWmq1JEE">

Required checks are picked manually in the repo settings, to which only the owner has access. Since the jobs evolved (matrix has different params now), old ones need to be deleted and the new ones selected. We shall do it once we have confirmation that everything's working as intended.

csagataj2 commented:
> @csagataj2 You were the one? Please share some good news with us and tell us if it builds.

It works as intended

qwrtln commented:
> > @csagataj2 You were the one? Please share some good news with us and tell us if it builds.
> 
> It works as intended

I'm glad we're leaving this _window_ of opportunity open for everyone

qwrtln commented:
@Heegu-sama We need your help here. Those yellow required checks which won't run now need to go away. They should be replaced by those which run successfully, but are not labelled as required.
